Second iteration

MarijaBorisova opened the pull request:
Please, make my code review on the second iteration

MarijaBorisovaAccenture committed:
Deleted the previos part of code and changes have been done according to Mentor's request: manually input of code and random separately. Manually input- 1 iteration finished, Random- without logic, only print

MarijaBorisovaAccenture committed:
Random number Console 1 iteration is finished

MarijaBorisovaAccenture committed:
Deleted the first ex before Field 0

MarijaBorisovaAccenture committed:
the task: clean every iteration un input print speed

MarijaBorisovaAccenture committed:
Some changes have been done

MarijaBorisovaAccenture committed:
Deleted unnecessary projects

MarijaBorisovaAccenture committed:
Changes have been done taking an account mentor's comments

MarijaBorisovaAccenture committed:
RandomConsole some changes

MarijaBorisovaAccenture committed:
During the testing, the mistake in NeighboursCount has been found. Resolved

MarijaBorisovaAccenture committed:
Changes have been done acc. to the mentor's requirements in pull requests. Selector has not been done.

MarijaBorisovaAccenture committed:
Switch selector on manual and Random input

MarijaBorisovaAccenture committed:
Acc to mentor's review two identical code parts have been input in one.

MarijaBorisovaAccenture committed:
Bug ir recognized and resolved: no clear action, everytime should be rerun. Resolved. Another symbol input- the error is shown. Random variant also works properly.

MarijaBorisovaAccenture committed:
Add private modifiers to the fields in GameSeed class.
Started the second interation: count of iteration added, file saving and restore.

MarijaBorisovaAccenture committed:
Second Iteration has been fulfilled:
count of iterations and alive cells, save data in a file, choice to stop app.

EdgarsBondars commented:
You have added multiple projects again.
The goal is to have one project that has all iteration functionalities.
For example there should be a new class within your project that has logic for writing to a file and reading from a file.
Iteration 2 requirement is that it should be possible to save a game into a file and have possibility not only to print saved game but to continue iterating through saved game.

EdgarsBondars commented:
There are plenty of materials on serialization/deserialization into a file.
For example https://docs.microsoft.com/en-us/dotnet/api/system.serializableattribute?view=net-6.0

MarijaBorisova commented:
Acc. to above mentioned comments, the following changes have been done:
the one project
possibility to save a game into a file and have possibility not only to print saved game but to continue iterating through saved game.

MarijaBorisovaAccenture committed:
the one project
possibility to save a game into a file and have possibility not only to print saved game but to continue iterating through saved game.

EdgarsBondars commented on GameOfLifeConsole/Services/FileReadSave.cs:
```diff
@@ -0,0 +1,66 @@
+            }
+        }
+
+        public void SaveDataR(GameSeed gameSeedRandom)
```
SaveDataR is same exact method as SaveData. Remove one of them.

MarijaBorisova replied:
Done. Thank you.

EdgarsBondars commented on GameOfLifeConsole/Services/FileReadSave.cs:
```diff
@@ -0,0 +1,66 @@
+namespace GameOfLifeConsole.Services
+{
+    /// <summary>
+    /// To read and save the data on HD.
```
To read and save the data on **HD**.

What is HD? Try to avoid any obviations in code comments or when naming new variables.

MarijaBorisova replied:
Done.

EdgarsBondars commented on GameOfLifeConsole/Services/FileReadSave.cs:
```diff
@@ -0,0 +1,66 @@
+    /// </summary>
+    class FileReadSave
+    {
+        // The path to which the data will be saved and restored. The variable
```
// The path to which the data will be saved and restored. **_The variable_**

Unfinished comment ?

MarijaBorisova replied:
Forgot to finish the comment with the "point-."

EdgarsBondars commented on GameOfLifeConsole/Services/FileReadSave.cs:
```diff
@@ -0,0 +1,66 @@
+    class FileReadSave
+    {
+        // The path to which the data will be saved and restored. The variable
+        // The path for FileReadSave object, ctor.
```
// The path for FileReadSave object, ctor.

Most likely this is path to some kind of game storage file not FileReadSave object. And why there is ctor mentioned I can't even imagine :D

MarijaBorisova replied:
Yeap, the wrong comment. Removed.

EdgarsBondars commented on GameOfLifeConsole/Services/FileReadSave.cs:
```diff
@@ -0,0 +1,66 @@
+        /// <summary>
+        /// Save data on HD.
+        /// </summary>
+        public void SaveData(GameSeed gameSeed)
```
Input parameter comment missing.

MarijaBorisova replied:
Done. Changed the name acc. to changed the class (GameSeed has been changed onto GameLogic)

EdgarsBondars commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+﻿using GameOfLifeConsole.Services;
```
Program.cs should be a file where we initialize game and tell it to start. It should be around 5 lines long. Here shouldn't be any logic. Create new class for all the UI related stuff. And new class where different patterns are stored/generated there you can put blinker/random generation.

MarijaBorisova replied:
Done.

EdgarsBondars commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+                            }
+                            _fileReadSave.SaveData(gameSeed);
+                            gameSeed = _fileReadSave.LoadData();
+                            try
```
What is this try catch it has no point.

MarijaBorisova replied:
Removed.

EdgarsBondars commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+                                }
+
+                                //in order to create the object, ctor of this class, path should be
+                                FileReadSave _fileReadSaveR = new FileReadSave();
```
Why not just use same _fileReadSave but create new one?

MarijaBorisova replied:
Removed. Agreed with you and leave the one.

EdgarsBondars commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+                                        {
+
+                                        }
+                                        Console.WriteLine();
```
Does this Console.WriteLine(); actually serve some purpose ?
for cycle at line 102 definitely can be removed, for cycle at line 100 too ? depends if this WriteLine actually has use.

MarijaBorisova replied:
Yes, agree. Removed. Thank you.

EdgarsBondars commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+                try
+                {
+
+                    uint MaxRuns = 20;
```
MaxRuns and runs don't belong here it makes sense to put them where game logic is.

MarijaBorisova replied:
Done.

GuntarsD commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+    class Program
+    {
+        static void Main(string[] args)
+        {
```
Main method should be small like:
![image](https://user-images.githubusercontent.com/103197809/165082739-d57640e1-c6f3-4178-8503-f7fc8bffdfe0.png)

MarijaBorisova replied:
Resolved.

GuntarsD commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+            while (true)
+            {
+                Console.Clear();
+                try
```
Try/catch is slow and should be used only in cases when there is no other means to ensure correct execution of code.

MarijaBorisova replied:
If I would remove try catch in this case, I will have the error and wrong run.

MarijaBorisova replied:
resolved.

GuntarsD commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+                        "LifeCircle with Random numbers (R). \t");
+                    Console.Write("Select : M or R:\t ");
+
+                    ConsoleKey consoleKey = Console.ReadKey().Key;
```
This and following line can be merged into one without creating variable consoleKey

MarijaBorisova replied:
Done.

GuntarsD commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+                    int runs = 0;
+                    int[,] seed = null;
+
+                    Console.WriteLine("App Title: Game Of Life");
```
Code used to communicate with user can be put into separate class dedicated for this purpose.

MarijaBorisova replied:
The Class AppUserInterface has been created to fulfill this goal.

GuntarsD commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+                    int[,] seed = null;
+
+                    Console.WriteLine("App Title: Game Of Life");
+                    Console.WriteLine("Please, select if you would like to see the Blinker life circle (M) or " +
```
Text literal should not be used directly in code.
Either make const variable at beginning of file or preferably create new class to store application constants.

MarijaBorisova replied:
Done.

GuntarsD replied:
You still have texts directly in the code.

MarijaBorisova replied:
Done.

GuntarsD commented on GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,165 @@
+                    switch (consoleKey)
+                    {
+                        case ConsoleKey.M:
+                            seed = new int[,]
```
All code from switch case blocks should be moved into separate methods.

MarijaBorisova replied:
Also done.

GuntarsD replied:
Nope. Switch case for "R" is one method only, but for case "M" you have long implementation.
That makes Run() method too long. Take code from  case "M" out into new method.

MarijaBorisova replied:
Done

GuntarsD commented on GameOfLifeConsole/Services/FileReadSave.cs:
```diff
@@ -0,0 +1,66 @@
+        public GameSeed LoadData()
+        {
+            // To load data from the file, need to check if the file exists.
+            var fileIs = File.Exists(Path);
```
Use var only when it is clear what type it is.
Like:
Var gameSeed = new GameSeed();

In this case use explicit bool.

Names of Boolean types should start with is, has , and so on.
Make it like doesFileExist

MarijaBorisova replied:
Done acc.to recommendations.

GuntarsD commented on GameOfLifeConsole/Services/FileReadSave.cs:
```diff
@@ -0,0 +1,66 @@
+                                                 // and free the resourses with Dispose method.
+
+                    return new GameSeed(gameField);
+                    //return null;
```
Remove obsolete text.

MarijaBorisova replied:
Done. Thank you.

GuntarsD commented on GameOfLifeConsole/Services/FileReadSave.cs:
```diff
@@ -0,0 +1,66 @@
+        /// </summary>
+        public void SaveData(GameSeed gameSeed)
+        {
+            // Using object StreamWriter which create taking the method CreateText at File class
```
Please end all comments with dot.

MarijaBorisova replied:
Done.

GuntarsD commented on GameOfLifeConsole/GameSeed.cs:
```diff
@@ -0,0 +1,123 @@
+        // Current field of cells, playfield, an array in which all the cells will be counted
+        // (alive or dead).
+        public int[,] gameField;
+        // Current field will be changed after new requirements.
```
Here in all other places add empty line before comment

MarijaBorisova replied:
Did it.

GuntarsD commented on GameOfLifeConsole/GameSeed.cs:
```diff
@@ -0,0 +1,123 @@
+    /// Logic of game.
+    /// </summary>
+   [Serializable]
+    public class GameSeed
```
Name of class looks more like GameData or GameLogic
It is suggestion, but consider to rename it.

MarijaBorisova replied:
Done acc.to recommendations.

GuntarsD commented on GameOfLifeConsole/GameSeed.cs:
```diff
@@ -0,0 +1,123 @@
+        /// <summary>
+        /// Print the current field on console.
+        /// </summary>
+        public void DrawField()
```
Drawing preferably should be put into separate class for tasks related to rendering.
For rendering data, displaying menu and retrieving data from user.

GuntarsD commented on GameOfLifeConsole/GameSeed.cs:
```diff
@@ -0,0 +1,123 @@
+                {
+                    if (!((i < 0 || j < 0) || (i >= column || j >= row)))
+                    {
+                        if (gameField[i, j] == 1)
```
In professional code always use block even there is only one statement.
In this particular case you don't need if.
Use:
count += gameField[i, j];

GuntarsD commented on GameOfLifeConsole/GameSeed.cs:
```diff
@@ -0,0 +1,123 @@
+            {
+                for (int j = y - 1; j < y + 1; j++)
+                {
+                    if (!((i < 0 || j < 0) || (i >= column || j >= row)))
```
Internal parentheses aren't need.

MarijaBorisova replied:
done.

GuntarsD commented on GameOfLifeConsole/GameSeed.cs:
```diff
@@ -0,0 +1,123 @@
+        {
+            int[,] newGameField = new int[column, row];
+            changedField = (int[,])gameField.Clone();
+
```
Empty line aren't needed.

GuntarsD commented on GameOfLifeConsole/GameSeed.cs:
```diff
@@ -0,0 +1,123 @@
+                        newGameField[y, x] = 0;
+                }
+            }
+            gameField = (int[,])newGameField.Clone();
```
Empty line needed after closing block not followed by other closing }.

MarijaBorisova replied:
done

MarijaBorisovaAccenture committed:
File restore without errors. The changes in the naming of classes: GameSeed on GameLogic, the new class has been added in which the UI has been transferred from Program. Program.cs is free from UI.

MarijaBorisovaAccenture committed:
Game class has been changed acc.mentor's comments. AppUI is added with additional methods.

MarijaBorisovaAccenture committed:
Changes in the FileReadSave in naming acc. to changed class name (GameSeed onto GameLogic)

EdgarsBondars commented on GameOfLifeConsole/AppUserInterface.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <returns> The visual part the message-request.</returns>
+        public static int RequestNumberFromUser(string promptMessage)
+        {
+            while (true)
```
This while true is kinda weird.
I'd do it like 
```
while(!int.TryParse(Console.ReadLine(), out int result))
{
// Here goes exception handling/ message printing  when input was wrong.
}
return result
````

MarijaBorisova replied:
I got it. But if it could be possible to leave as it is? in this variant it shows me the error.

EdgarsBondars commented on GameOfLifeConsole/AppUserInterface.cs:
```diff
@@ -0,0 +1,86 @@
+                string value = Console.ReadLine();
+                if (int.TryParse(value, out int result))
+                {
+                    return result;
```
Variable name result tells us nothing atleast rename it to userInput or userIntegerInput.

MarijaBorisova replied:
Done. Thank you.

EdgarsBondars commented on GameOfLifeConsole/AppUserInterface.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <summary>
+        /// The method if the user input data incorrectly.
+        /// </summary>
+        public static void IncorrectDataInput()
```
Create new file with messages that will be outputed to user.

MarijaBorisova replied:
Done as Repository class.

EdgarsBondars commented on GameOfLifeConsole/AppUserInterface.cs:
```diff
@@ -0,0 +1,86 @@
+            Console.WriteLine("\nInput incorrect data! Please, try again.");
+        }
+
+        public static void StopIteration()
```
This has 0 references.
I see there is same exact code but with break you can keep this method but it should return bool.
so if stop == s return true and by default return false.
And in Game.cs
``` 
if(AppUserInterface.StopIteration())
{
    break;
}
````

MarijaBorisova replied:
This method has been removed.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,133 @@
+    {
+        private GameLogic gameLogic;
+        FileReadSave _fileReadSave = new FileReadSave();
+        uint MaxRuns = 20;
```
Access modifiers missing for parameters.

MarijaBorisova replied:
Done. Added.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,133 @@
+                string key;
+                key=Console.ReadLine();
+                //ConsoleKey consoleKey = Console.ReadKey().Key;
+                switch (key)
```
You can read key like 
```
switch (Console.ReadLine())
````
and remove commented out code above/below.

MarijaBorisova replied:
Done. Thank you.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,133 @@
+                AppUserInterface.Start();
+                string key;
+                key=Console.ReadLine();
+                //ConsoleKey consoleKey = Console.ReadKey().Key;
```
Why you didn't like using ConsoleKey ?

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,133 @@
+                    {0,0,0,0,0},
+                    {0,1,1,1,0},
+                    {0,0,0,0,0},
+                    {0,0,0,0,0}
```
Create class for game presets and put this initiation there.
```
public int[,] GenerateBlinker 
{
    return new int[,]
    {
         {0,0,0,0,0},
         {0,0,0,0,0},
         {0,1,1,1,0},
         {0,0,0,0,0},
         {0,0,0,0,0}
    }
}
````
Do same with random where it gets array and returns array with randomly generated cells.

MarijaBorisova replied:
Resolved in the separate class acc.to recommendations.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,133 @@
+                }
+                gameLogic.AliveCells().ToString();
+
+                while (gameLogic.AliveCells() > 0 && runs++ < MaxRuns)
```
Again you have this huge while loop repeating twice in this file we had sorted it out before. Put it after switch so it doesn't repeat.

MarijaBorisova replied:
Resolved acc. to recommendations.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,133 @@
+
+                    default:
+
+                        AppUserInterface.IncorrectDataInput();
```
Put curly brackets after switch statements
```
case "R"
{
// code here
}
default:
{
// code here
}
````
It will be easier to orient through code.

MarijaBorisova replied:
Done.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,133 @@
+                }
+            }
+        }
+        public void GameValidation()
```
This is called GameValidation but here it iterates game?

MarijaBorisova replied:
Renamed.

EdgarsBondars commented:
You misunderstood what we ment by implementing MVC pattern, you don't have to create a web application. You gotta apply MVC concepts to your console app. So please don't waste time on web app for now.

MarijaBorisovaAccenture committed:
Still the problem with restore data during running is.

GuntarsD commented on GameOfLifeConsole/GameLogic.cs:
```diff
@@ -0,0 +1,139 @@
+                    }
+                    //Any live cell with two or three live neighbours lives on to the next generation.
+                    else if (gameField[y, x] == 1 &&
+                           (neighboursNumber == 2 || neighboursNumber == 3))
```
Wrong alignment

MarijaBorisova replied:
Removed.

MarijaBorisovaAccenture committed:
GameLogic is working properly.

MarijaBorisovaAccenture committed:
App could save during its running.

MarijaBorisovaAccenture committed:
2nd iteration finished.

GuntarsD commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,103 @@
+
+namespace GameOfLifeConsole
+{
+    public class Game
```
Please add comments for class and methods.

MarijaBorisova replied:
Done. Thank you.

GuntarsD commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+            gameLogic = new GameLogic();
+        }
+        public void Run()
```
Missing empty line between methods

MarijaBorisova replied:
Done. Thank you.

GuntarsD commented on GameOfLifeMVC/Models/GameModel.cs:
```diff
@@ -0,0 +1,38 @@
+        // Current field of cells, playfield, an array in which all the cells will be counted
+        // (alive or dead).
+        public int[,] gameField;
+        // Current field will be changed after new requirements.
```
Add empty line before comment

MarijaBorisova replied:
The MVC project has been deleted.

GuntarsD commented on GameOfLifeMVC/Models/GameModel.cs:
```diff
@@ -0,0 +1,38 @@
+        }
+
+        /// <summary>
+        /// Const. creates the field of cells by cloning the array.
```
Add class comment. Comment for constructor is the same.

MarijaBorisova replied:
MVC project has been removed.

MarijaBorisovaAccenture committed:
2nd iteration without MVC project adding

MarijaBorisovaAccenture committed:
Fulfilled the requerements acc.mentor's comments.

EdgarsBondars commented on GameOfLifeConsole/AppUserInterface.cs:
```diff
@@ -0,0 +1,71 @@
+            Console.WriteLine("App Title: Game Of Life");
+            Console.WriteLine("Please, select if you would like to see the Blinker life circle (M) or " +
+                "LifeCircle with Random numbers (R). \t");
+            Console.Write("Select : M or R:\t ");   
```
These can be combined into one string and printed in one go.
Put these to repository file aswell.

MarijaBorisova replied:
Done. Thanks.

EdgarsBondars commented on GameOfLifeConsole/FieldGeneration.cs:
```diff
@@ -0,0 +1,48 @@
+            var gameField = new int[columnsInField, rowsInField];
+            Random random = new Random();
+
+            for (int row = 0; row < gameField.GetLength(1); row++) // Solid principles, the code of random numbers generating.
```
This comment makes no sense to me, this for cycle has nothing to do with solid principles and second part is unnecessary as summary comment tells us enough.
I'd say remove this coment.

MarijaBorisova replied:
Done. Agree.

EdgarsBondars commented on GameOfLifeConsole/FieldGeneration.cs:
```diff
@@ -0,0 +1,48 @@
+﻿using GameOfLifeConsole.Services;
```
Unnecessary using, run visual studio code cleanup.

MarijaBorisova replied:
Done.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,116 @@
+                while (gameLogic.AliveCells() > 0 && runs++ < MaxRuns)
+                {
+                    string exit;
+                    string save;
```
Initialize variables where they are used, in this case initialize save on line 71 and exit on line 78.

MarijaBorisova replied:
Done.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,116 @@
+                    string exit;
+                    string save;
+                    Console.Clear();
+                    Console.Title = gameLogic.countIteration.ToString("Iteration {0}");
```
You are missing alive cell counter. Add it either in console title or anywhere you like it.

MarijaBorisova replied:
I have alive cell counter in class "GameLogic" on line 68.
 Console.WriteLine("Alive cells number: " + count);
Does not it correct?

EdgarsBondars replied:
Ah I didn't notice. Currently you have it on pause screen. Would be nice to have opportunity to see the count when game runs in real time, every second it would print alive cell count.

EdgarsBondars replied:
Oh I see you have it implemented but the problem is first it prints alive cell count but then you put cursor position on first line and it gets overwritten when you start printing the game.

MarijaBorisova replied:
Done. Alive Cells number input in the Title.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,116 @@
+                AppUserInterface.IncorrectDataInput();
+                Console.ReadLine();
+            }
+            gameLogic.AliveCells().ToString();
```
```
gameLogic.AliveCells().ToString();
````
This seems to do nothing useful to me. Can it be removed ? 
Same in StartGameManually.

MarijaBorisova replied:
It initiates counting of Alive Cells. If I would remove it- alive cells would not be counted.

EdgarsBondars commented on GameOfLifeConsole/GameLogic.cs:
```diff
@@ -0,0 +1,134 @@
+                {
+                    int neighboursNumber = NeighboursCount(row, column);
+                    // Any live cell with fewer than two live neighbours dies, as if by underpopulation.
+                    if (gameField[column, row] == 1 && neighboursNumber <= 2)
```
_Any live cell with two or three live neighbours lives on to the next generation._
I think what you have here is if cell has two neighbours it dies.
And then in next else if you mark that cell alive so you have mistake in code but it works as expected but still change it  :)

MarijaBorisova replied:
Removed '=' sign.

EdgarsBondars commented on GameOfLifeConsole/Repository.cs:
```diff
@@ -0,0 +1,23 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Unnecessary usings.

MarijaBorisova replied:
Done.

EdgarsBondars commented on GameOfLifeConsole/Repository.cs:
```diff
@@ -0,0 +1,23 @@
+        public const string stopOrContinue = "\nTo continue iteration, press the key 'Enter'.\n" +
+                            " \nTo quit iteration, press 'e' key.";
+        public const string cellsDied = "Everyone is died!";
+        public const string saveData = "\nTo sava data in the file, press 's' key.\n";
```
public constant string names should be in PascalCase.

MarijaBorisova replied:
Done. Thank you.

EdgarsBondars commented on GameOfLifeConsole/Services/FileReadSave.cs:
```diff
@@ -0,0 +1,61 @@
+    {
+        // The path to which the data will be saved and restored. The variable.
+        private string Path = $"{Environment.CurrentDirectory}\\MyFile.json";
+        private int[,] gameField;
```
private properties should start with underscore and should continue with lower camelCase.

MarijaBorisova replied:
Got it. Thank you. Done.

EdgarsBondars commented on GameOfLifeConsole/Game.cs:
```diff
@@ -0,0 +1,116 @@
+        public GameLogic gameLogic;
+        private FileReadSave _fileReadSave = new FileReadSave();
+        private uint MaxRuns = 50;
+        private int runs = 0;
```
Private properties should start with underscore and should continue with lower camelCase.
public property name should be in upper CamelCase.

MarijaBorisova replied:
Thank you. Done.

MarijaBorisovaAccenture committed:
Changes made acc.mentor's comments.

MarijaBorisovaAccenture committed:
In GameLogic 109 line changes acc.rules

MarijaBorisovaAccenture committed:
Added Alive Cells number in the running game title.
